### PR TITLE
[Model] Add RECT example

### PR DESCRIPTION
**Pull request by Fizyhsp**
### **Description**

A proposal of a new GNN model coming from TKDE 2020: https://ieeexplore.ieee.org/document/8979355
The original implementation is: https://github.com/zhengwang100/RECT

This is initial proposal of the model. There are five files: `utils.py` which loads dataset and preprocesses node features, `label_utils.py` which removes the unseen classes and gets the class-center (semantic knowledge) of each seen class, `model.py` which implements GCN model and RECT-L model, `main.py` which shows training process, and `classify.py` which defines a logistic regression classifier to evaluate the quality of embedding results.

We use two DGL's build-in datasets (Cora and Citeseer) with their default train/val/test settings. The performance results are are as follows:

| **Datasets/Models** | **NodeFeats** | **GCN** | **RECT-L** |
| :-----------------: | :-----------: | :-----: | :--------: |
|      **Cora**       |     47.56     |  51.26  | **68.60**  |
|    **Citeseer**     |     42.04     |  37.55  | **56.32**  |

<center>Table 1：node classification results with the first three classes as "unseen"</center>

| **Datasets/Models** | **NodeFeats** | **GCN** | **RECT-L** |
| :-----------------: | :-----------: | :-----: | :--------: |
|      **Cora**       |     47.56     |  56.91  | **69.30**  |
|    **Citeseer**     |     42.04     |  45.69  | **61.85**  |

<center>Table 2：node classification results with the last two classes as "unseen"</center>

We think RECT should be part of the DGL, as the zero-shot label setting is very common in practical applications. On the other hand, existing methods (like GCN) perform very badly in this case. While our method outperforms GCN very significantly, around 22%-50% relatively.

We are happy to answer any questions and discuss how to integrate our method in DGL in the optimal way. Please feel free to contact us. 

Authors: Tingzhang Zhao (undergraduate, @USTB), and Zheng Wang (supervisor, the original author of this paper, AP@USTB). 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by BarclayII** [comment]
Is it possible to share the paper (at least privately on Slack)?  I cannot download it from the IEEE link above.

File: examples/pytorch/rect/classify.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from statistics import mean
+
+class logistic_regression_classifier(nn.Module):
```
Comment (BarclayII) [inline examples/pytorch/rect/classify.py:6]: We suggest using CamelCase for class names:
```
class LogisticRegressionClassifier(nn.Module):
    ...
```

File: examples/pytorch/rect/classify.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    for _ in range(100):
+        model.train()
+        optimizer.zero_grad()
+        output = model(features)
+        loss_train = F.cross_entropy(output[train_mask], labels[train_mask])
+        loss_train.backward()
+        optimizer.step()
+    return _evaluate(model=model, features=features, labels=labels, test_mask=test_mask)
+
+def evaluate_embedds(features, labels, train_mask, test_mask, n_classes, cuda, test_times=10):
```
Comment (BarclayII) [inline examples/pytorch/rect/classify.py:39]: `evaluate_embeds`?

**Review by BarclayII — APPROVED** [review]
I'm good.  @sneakerkg could you also have a pass on this PR?

File: examples/pytorch/rect/README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+
+## **Performance**
+
+The performance results are are as follows:
+
+| **Datasets/Models** | **NodeFeats** | **GCN** | **RECT-L** |
+| :-----------------: | :-----------: | :-----: | :--------: |
+|      **Cora**       |     47.56     |  51.26  | **68.60**  |
+|    **Citeseer**     |     42.04     |  37.55  | **56.32**  |
```
Comment (sneakerkg) [inline examples/pytorch/rect/README.md:30]: Which setting are we using to get the performance? Does it match with the paper performance?

File: examples/pytorch/rect/README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# **DGL Implementation of RECT (TKDE20)**
+
+This DGL example implements the GNN model **RECT** (or more specifically its supervised part **RECT-L**) proposed in the paper [Network Embedding with Completely-imbalanced Labels](https://ieeexplore.ieee.org/document/8979355). The authors' original implementation can be found [here](https://github.com/zhengwang100/RECT).
+
```
Comment (sneakerkg) [inline examples/pytorch/rect/README.md:4]: Thanks for contributing to DGL! Maybe you can add your github information in README like: This example was implemented by Fizyhsp.

File: examples/pytorch/rect/label_utils.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            label_attribute_nodes[int(label)].append(int(nodeid))
+    label_attribute = {}
+    for label in label_attribute_nodes.keys():
+        nodes = label_attribute_nodes[int(label)]
+        selected_features = features[nodes, :]
+        label_attribute[int(label)] = np.mean(selected_features, axis=0)
+    return label_attribute
+
+def get_labeled_nodes_label_attribute(train_mask_zs, labels, features, cuda):
+    ''' Replace the original labels by their class-centers.
```
Comment (sneakerkg) [inline examples/pytorch/rect/label_utils.py:48]: The comments are not aligned here for different rows.

File: examples/pytorch/rect/README.md (hunk 3 above)
Comment (Fizyhsp) [reply examples/pytorch/rect/README.md:30]: Thanks.

Firstly, as described in the Readme.md, we adopt the standard split of DGL datasets and adopt some (user-specified) classes as unseen. 

Secondly, in the original paper, it randomly selects some percents of labeled data and also randomly adopts some unseen classes.
Therefore, we do not get the same performance.

File: examples/pytorch/rect/classify.py (hunk 2 above)
Comment (Fizyhsp) [reply examples/pytorch/rect/classify.py]: @BarclayII I modified the class names and function names mentioned above.

File: examples/pytorch/rect/label_utils.py (hunk 5 above)
Comment (Fizyhsp) [reply examples/pytorch/rect/label_utils.py:47]: @sneakerkg I added my github information and also adjusted the comments.

**Review by sneakerkg — APPROVED** [review]

**Comment by sneakerkg** [comment]
@Fizyhsp We just merged the PR. Thanks for contributing to DGL!
